### ISSUE TITLE
Problem when running "flutter build aar" using AGP 8.1 + JDK 17

### ISSUE THREAD

**Issue by caveg002** [flutter/flutter#154371]
### Steps to reproduce

1. Set up a flutter module with Flutter 22.2 + AGP 8.1 + JDK 17
2. Try to generate the AAR running `flutter build aar`

Note: this project has dependency on other flutter plugins, published in dispub, as well as other 3rd party and Android AAR dependencies.

### Actual results

gives this error:
```
FAILURE: Build failed with an exception.

Where:
Initialization script '/Users/user/flutter_envs/flutter_macos_3.22.2-stable/flutter/packages/flutter_tools/gradle/aar_init_script.gradle' line: 148

What went wrong:
Cannot invoke method dependsOn() on null object. 

```

### Logs

<details open>
<summary>Logs</summary>

```console

<!-- Paste your logs here -->

```

</details>

### Flutter Doctor output

<details open>
<summary>Doctor output</summary>

```console

<!-- Paste your output here -->

```

</details>

**Comment by huycozy** [comment]
Hi @caveg002 

>Set up a flutter module with Flutter 22.2 + AGP 8.1 + JDK 17

Do you set AGP inside Android native host project in `build.gradle.kts` file? My Android host project is using AGP 8.1.1 and Gradle 8.2 but it builds aar fine. 

```kts
plugins {
    id("com.android.application") version "8.1.1" apply false
    id("org.jetbrains.kotlin.android") version "1.9.0" apply false
}
```

```properties
distributionUrl=https\://services.gradle.org/distributions/gradle-8.2-bin.zip
```

You can try my complete sample project here to see if it can build aar for Flutter module on your end: [Flutter-Add2App-Compose](https://github.com/huycozy/Flutter-Add2App-Compose). I'm on the latest Flutter stable version 3.24.1.

Please provide the output of `flutter doctor -v` as well. Thanks!

**Comment by LOPEJ670** [comment]
**Issue:** Error generating AAR in Flutter module after upgrading to AGP 8.1.4

**Description:**

We have a Flutter module that depends on other modules, both private and third-party. Up until AGP version 7.\*, the `build AAR` command worked as expected. However, after updating the libraries to version 8.1.4, including the parent dependency that contains all others, the build process started failing.

**Current Configuration:**

Within the `.android` folder, we modified the `build.gradle` as follows:

```groovy
buildscript {
    ext.kotlin_version = "1.8.20"
    repositories {
        google()
        mavenCentral()
    }

    dependencies {
        classpath("com.android.tools.build:gradle:8.1.4")
        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version")
    }
}
```

In the app’s `build.gradle` file:

```groovy
android {
    // Compatibility with AGP <4.2.
    if (project.android.hasProperty("namespace")) {
        namespace = "our namespace"
    }

    compileSdk = 34

    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_17
        targetCompatibility = JavaVersion.VERSION_17
    }
}
```

**Problem:**

With these changes, we started encountering the error mentioned previously. Upon investigation, we found that starting with AGP 8.0, it's necessary to explicitly configure the `publish variant` since AGP 8.0 no longer creates any `SoftwareComponent` by default. More details here: [Configure Publish Variants](https://developer.android.com/build/publish-library/configure-pub-variants#groovy) and [AGP 8.0 Release Notes](https://developer.android.com/build/releases/past-releases/agp-8-0-0-release-notes#enforced-changes).

We tried using the flag `android.disableAutomaticComponentCreation=false`, but it didn’t work as this was deprecated in AGP 8.1. We tested this in versions 3.22.2 and 3.24.1, both locally and on the server, with the correct configuration in both environments. Additionally, we added the publish variant within the `.android/Flutter/build.gradle` file, without success.

**Reproduction of the Issue:**

To reproduce this issue, you can create a new Flutter module using:

```bash
flutter create -t module --org com.example flutter_module
```

Then, update to AGP 8.1.4 and Kotlin 1.8.20. We added the third-party plugin `video_player_android: ^2.6.0` (and others) and encountered an error in `aar_init_script.gradle`.


If you need more details or have any suggestions on how to resolve this issue, any help would be greatly appreciated.

**Comment by LOPEJ670** [comment]
Hi again!

We have modified the code in aar_init_script.gradle to allow for successful AAR generation using AGP 8.+. This change is necessary, and it also requires including the publish variants within the _.android/Flutter/build.gradle_ file. To make the review easier, here is a link to a draft where you can see the changes. We hope this helps or can be optimized. https://github.com/flutter/flutter/pull/154455

Thanks!

**Comment by gmackall** [comment]
This is part of the way towards being fixed (but not fixed) on master currently.

We recently bumped [module templates to use AGP 8.1](https://github.com/flutter/flutter/pull/151675), and noticed that it broke the `aar_init_script.gradle` that adds the necessary gradle task for publishing the aar. This is because AGP 8.0+ doesn't publish software components by default, and we [rely on these components existing](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/aar_init_script.gradle#L29) in the init script.

We tried to work around this by using AGP's publishing dsl to re-create these software components, see these lines:
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl#L50-L64

But this doesn't take in to account that we need these components to exist for the plugin projects as well, which explains why @huycozy initially found the module to build correctly, but then @LOPEJ670 pointed out that it breaks when you add a plugin. 

Looking in to if we can do a similar thing to what we added to the ephemeral module template (aka, the [earlier link](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl#L50-L64)), but within the init script, and applied to each plugin project.

**Comment by huycozy** [comment]
Thanks @LOPEJ670, I can also reproduce this issue now after updating AGP version on `.android` folder and adding a plugin `video_player_android`. The complete sample project can be found at [agp8-androidmodule-154371](https://github.com/huycozy/Flutter-Add2App-Compose/tree/agp8-androidmodule-154371). 

Reproduced on Flutter stable `3.24.2` and master `3.25.0-1.0.pre.283`.

**Comment by gmackall** [comment]
Still looking in to this, latest is that despite using the publishing DSL to try to publish some variants on the plugin projects (to re-create the software components), the components still aren't getting created. Weirdly, when logging the variants show when referenced in the modules gradle files:

```
IN PROJECT video_player_android (to be exact, project ':video_player_android')
IT has the android extension, and singlevariants of size 2
```

But not when referenced from the init script:
```
Plugin is video_player_android (to be exact, plugin is project ':video_player_android')
Plugin has software components: [] and singlevariants 0
```

Which suggests that perhaps they are being overridden somewhere in between - looking in to that at the moment

**Comment by gmackall** [comment]
Confirmed that the aar builds successfully if we add 
```
    buildTypes.all {buildType ->
        publishing.singleVariant(buildType.name) {
            withSourcesJar()
            withJavadocJar()
        }
    }
```
To `video_player_android`'s build.grade, which is good news in that the same fix that works for the module itself should also work for plugins. Still trying to figure out why trying to define the same variants (silently) fails when we do it in either the init script, or the [Flutter Gradle plugin](https://github.com/flutter/flutter/blob/6ad6641f0f27bd3c698a6957532971068f1bc2c5/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L137) (trying to avoid having the answer be "every plugin needs to update their gradle files").

**Comment by LOPEJ670** [comment]
Hi @gmackall,

I was just about to mention that, in the case of libraries that don’t have the singleVariants configuration defined, you can add the following code in .android/build.gradle to avoid having to manually adjust each library:

subprojects {
  apply plugin: 'maven-publish'

  afterEvaluate {
    if (!plugins.hasPlugin("android")) {
      publishing {
        publications {
          release(MavenPublication) {
            from components.findByName('release')
            groupId = project.group
            artifactId = "${project.name}_release"
            version = project.version
          }

          debug(MavenPublication) {
            from components.findByName('debug')
            groupId = project.group
            artifactId = "${project.name}_debug"
            version = project.version
          }
        }
      }
    }
  }
}

With this, the component now exists and can be generated. You only need to ensure that in the aar_init_script, the existence of the publish task is checked to prevent another error. For now, it takes the version from the same project.

**Comment by gmackall** [comment]
@LOPEJ670, thanks, yep I ended up doing something similar in this work-in-progress PR https://github.com/flutter/flutter/pull/154757 (but in the init script instead of in the `.android/build.gradle`. 

Still testing it at the moment!

**Comment by caveg002** [comment]
Hi [gmackall](https://github.com/gmackall), 
Do you have an ETA on when this will be fixed and included in Master? Any idea what version will it be ?

Thanks

**Comment by gmackall** [comment]
> Hi [gmackall](https://github.com/gmackall), Do you have an ETA on when this will be fixed and included in Master? Any idea what version will it be ?
> 
> Thanks

I expect to merge the fix tomorrow, unless I run into unexpected complications. If that goes smoothly, I should be able to [cherry pick](https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-Cherrypick-Process.md) the PR (along with https://github.com/flutter/flutter/pull/151675), and get it included in `3.24.3`.

**Comment by gmackall** [comment]
Re-opening for now, just to track the work of landing the cherry picks (the problem should be fixed on master already).

**Comment by gmackall** [comment]
The relevant cherry pick (https://github.com/flutter/flutter/pull/154944) has landed, and I've verified that building an aar (from a module that uses a plugin) works as expected with AGP 8.1.

**Comment by gmackall** [comment]
The fix should now be available on `3.24.3`

**Comment by LOPEJ670** [comment]
Thanks @gmackall and @huycozy.
